### DISCUSSION TITLE
Change Room Description for Characters

### DISCUSSION THREAD

**Discussion by RantarR** [evennia/evennia#2565] [Evennia Questions] [answered]
Hello!

I have two problems with the default look command that unfortunately I could not solve myself. 

1) What function do i have to overwrite in order to change how the room displays characters?
Right now its like this: "You see: a wolf"
I would like to get rid of this line and instead display a short describtion that is saved on the character.
(for example: "A wolf stands here and growls at you.")

2) How do i rename the desc-attribute? (I would like to change it to a German term)

**Comment by Griatch** [comment, marked as answer]
Rather than just give the answer, let's see how one can figure something like this out from the code. The trick to understanding what goes on by default is to trace the code backwards. For this example: 

- You know that the `look` command will cause the `"You see: a wolf"` text to appear. So we need to figure out how the `look` command does it. Let's find the default `look` command in code. 
- You can use `git grep` in the evennia code or you can search documentation for "look". You will find [CmdLook](https://github.com/evennia/evennia/blob/master/evennia/commands/default/general.py#L55).
- Looking at the source of `CmdLook` you'll find that it contains this line `desc = caller.at_look(target)`. That is, it is calling a method `at_look` on the `caller`. The `caller` is an `Object` in this case - a `Character` more precisely. So let's look up the default `Character` typeclass.
- The `Character` typeclass is in your game folder, in `mygame/typeclasses/characters.py`. If you didn't change it, you'll find it's empty and inherits from `DefaultCharacter`, which is imported from `evennia`. 
- You can look in `evennia/__init__.py` to find that `DefaultCharacter` is just a shortcut to `evennia.objects.objects.Character`. So let's look there. You can also just search the docs for `DefaultCharacter` to find the class. 
- [DefaultCharacter](https://github.com/evennia/evennia/blob/master/evennia/objects/objects.py#L2033) doesn't on its own define any `at_look` method. So it must be defined in its parent class - in `DefaultObject`. That's in the same module.
- If you look at the code for [DefaultObject.at_look](https://github.com/evennia/evennia/blob/master/evennia/objects/objects.py#L1705) you'll see that it does some look-related permission checks and then the line `description = target.return_appearance(self, **kwargs)`. Here `target` is the thing you are looking at. So each object can (if they want) change how they are described! We'll find `return_appearance` on the same `DefaultObject` class too (all `Objects` can be looked at, after all). 
- [DefaultObject.return_appearance](https://github.com/evennia/evennia/blob/master/evennia/objects/objects.py#L1657) is the thing you want to modify. It will build the display string, including the "You see:" text. It will also access the `desc` Attribute. 

To override this, copy the `return_appearance` code into your own `mygame.typeclasses.characters.Character` class and edit it as you like. Reload to make use of your new variation. 

You can also have it use another attribute than `desc` if you want ... But honestly, as a non-native English speaker myself, I _strongly_ advice against you translating your _code_ into another language. 
Since Python itself is in English, mixing just makes your code hard to read (I've read a lot of code written by Germans and naming your variables in German is a really, really bad practice you should stop immediately). 
So I would recommend translating _user-facing strings_ to German, but to keep all your code variables in the original English.

**Reply by RantarR** [reply]
Thanks again Griatch for your detailed answer. I tried before to track the command back as you did but only came to the 
desc = caller.at_look(target) part and couldnt find it in the DefaultCharacter file.

Regarding the translation part, i guess you are right. My intention was it to make it easier for builders by using only german attributes. But that doesn't seem to be worth the effort so i will not translate the default code attributes like "description".

**Reply by RantarR** [reply]
After several hours of fiddling, I have to set one thing right for noobs like me.
In order to change how objects are displayed with the "look" command, you have to change the at_look(target) 
method of **rooms.py** and not characters.py. 
Characters.py will only work when you look direclty at characters but wont get called when you just type "look".

**Reply by Griatch** [reply]
That doesn't sound right. `at_look` is called on the `caller`, which is the `Character` firing the `look` command. It's `return_appearance` which is called on the thing you are looking at. Make sure you are puppeting a Character when you look - there's a separate OOC-look (and another one before you log in), but when OOC you cannot look at anything in the game world anyway, so you want to look at the regular in-character CmdLook.

The calling sequence is 

```python
# in CmdLook (evennia.commands.default.general.CmdLook)
desc = caller.at_look(target)
caller.msg(desc)   # essentially - this is what shows the result to the caller

# In caller of look-command (Character in this case)
def at_look(self, target):
    return target.return_appearance(looker=self)

# In target entity being looked at (Room in this case)
def return_apperance(self, looker):
    # builds the description of itself
    return desc
```

**Reply by RantarR** [reply]
Sorry, you are right. I meant the "return_appearance" Method, not the "at_look" Method. It worked for me after i override the "return_appearance" of typeclasses/rooms.py.

**Reply by Griatch** [reply]
Yes, that is correct. The `return_appearance` allows an object to describe itself.

By default, Characters and Rooms use the same `return_appearance` since they are both inheriting from `DefaultObject` which is why I pointed to that. If you want rooms only to have your change then it's correct to put your change only in the Room class as you did. If you want your change to apply to all objects (rooms/exits/characters alike), you should put the change in a class that all of them inherit from. 
For example, you could make `mygame.typeclasses.objects.Object` be that by adding your `return_appearance` to that class and then change all the other typeclasses to inherit both from that and from their base class: 

```python
# in typeclasses/objects.py
class Object(DefaultObject):
    # put your custom return_apperance here

```python
# in typeclasses/rooms.py (and same concept in exits.py and characters.py)
from typeclasses.objects import Object 
class Room(Object, DefaultRoom):   # <- changed to Multiple inheritance, first getting stuff from Object, then from DefaultRoom
     pass 

```
Any change you now do in `Object` will apply the same to your rooms, characters and exits.
(This is not anything Evennia-specific btw, but the way Python multi-inheritance works).

**Comment by dimitrilw** [comment]
Having done dev work in a multi-national setting, I'd say you really want to retain English as the core to your code. However, I also fully understand wanting to have some easy terms for your builders. In my own past, I've used getter/setter aliases as a workaround. However, I don't really encourage this. I'll add why, but first, here's a rough example in a proto-class:
```
class SomeThing:
    """
    docstring
    """
    def __init__(self, description: str = "default"):
        """ constructor """
        self.description = description
    
    @property
    def bezeichnung(self):
        """ Deutsche getter """
        return self.description
    
    @bezeichnung.setter
    def bezeichnung(self, description):
        """ Deutsche setter """
        self.description = description
```

The reasons I don't suggest doing this.

First, you don't have tests for these getter/setter properties. So, go make those too. Next, you will find yourself (or other devs) tempted to use these properties instead of the English ones, meaning you are highly likely to introduce inconsistent coding practices. Also, the actual var is still "description" so if the object is inspected, the term "bezeichnung" is not present in the `SomeThing.__dict__` nor in the DB. That can be confusing. ....best to stick with unaliases English variable names.

However, if you can avoid temptation and *only* use them as processing aliases to help your in-game builders, then the above is a work-around.

Cheers.

**Reply by dimitrilw** [reply]
I should add that I never did this workaround in the context of Evennia, but rather to totally unrelated python project. Same pitfalls apply, I would imagine.

**Reply by RantarR** [reply]
Do you also use only english names for Charakter-Stats?
For example:
"Stärke = " char.db.strength

**Reply by dimitrilw** [reply]
I didn't do this in context of a game. It was a special data analysis processor used by NATO personnel. However, we had some key demands for French terms. It was a whole thing. Anyway, we also aliased methods:
```
def do_thing(arg1, arg2):
     return arg1 + arg2
def french_do_thing(*args, **kwargs):
    return do_thing(*args, **kwargs)
```
That's rough-code (just tapped from eyesight), but you get the idea. It's a pass-through function. You can do the same with object methods ( `def do_thing(self, arg1, arg2)` & `def french_do_thing(self...` ), but again, the same pitfalls exist. And, errors get really rough to parse. Your tracebacks turn into an extra-snarled mess. Really don't recommend doing this *unless* you have a paying customer (NATO) who demands it and offers cash. ......I'm guessing your MUD isn't paying you, so trust me when I say that it introduces more problems than it fixes. 

Better to expand your software's help files & docstrings with extensive translation guides or something.

**Comment by Griatch** [comment]
Using the real variable name as a parameter in your translated string is not too bad and also makes it very clear where the boundary of forward-facing and backend code is:

    f"Styrka: {strength}, Magi: {magic}, Uthållighet: {stamina}"

**Comment by RantarR** [comment]
Thank you very much for your advice. As you probably noticed, I am only a hobby programmer and have never programmed in a professional context. However, I try to stick to the programming conventions as much as possible and have now changed my code to English.
